### ISSUE TITLE
Bugs in CultureInfo.GetCultures on Windows

### ISSUE THREAD

**Issue by mklement0** [dotnet/corefx#32207]
Observed on v2.1.401 (`dotnet --version`) on _Windows_.

Note that reproducing some of the issues requires the creation of custom locales via the [Locale Builder utility](https://www.microsoft.com/en-us/download/details.aspx?id=41158).

* `CultureInfo.GetCultures(CultureTypes.UserCustomCulture`) and  `CultureInfo.GetCultures(CultureTypes.ReplacementCultures)`  seemingly report ALL predefined cultures (same as `CultureInfo.GetCultures(CultureTypes.AllCultures)`).
With no custom locales installed (the default), NO cultures should be reported.

* A `CultureInfo` instance representing a _replacement_ custom culture does not have the `UserCustomCulture` flag set in its `.CultureTypes` property.
To see this problem, you must create and install a replacement custom locale with  Locale Builder.

* _All_ `CultureInfo` instances unexpectedly have the `InstalledWin32Cultures` flag set in `.CultureTypes`, even ad-hoc-created custom cultures, verify with `new CultureInfo("zz-YY").CultureTypes`

I haven't looked into behavior on Unix-like platforms.

**Comment by mklement0** [comment]
First - as a separate issue - let me say that the documentation needs updating: https://docs.microsoft.com/en-us/dotnet/api/system.globalization.culturetypes?view=netcore-2.1 (and related documentation):

* only deals with Windows
* only mentions `FrameworkCultures` and `WindowsOnlyCultures` as deprecated, neither of which are the subject of this bug report.

---

> avoid using it especially these will not make sense anyway on Linux/Mac.

* So, to be clear, you're saying that culture types `CultureTypes.UserCustomCulture` and `CultureTypes.ReplacementCultures` are deprecated in CoreFx _even on Windows, where they still make sense_?

Note that we're not just talking about passing those flags to `.GetCultures()`, but also what the individual `CultureInfo` instances report in their `.CultureTypes` property, so you can infer their specific type.

* And you're saying that on Unix-like platforms the concepts of supplemental and replacement cultures (collectively: _custom_ cultures) _do not apply altogether_?

Note that `UserCustomCulture` _is_ being meaningfully applied in a meaningful manner even on macOS / Linux, in that a `CultureInfo` instance created ad-hoc (one that is not predefined) does contain that flag in its `.CultureTypes` property - compare the following on Unix (using PowerShell):

```powershell
# Predefined culture.
PS> [CultureInfo]::new("en-US").CultureTypes
SpecificCultures

# Ad-hoc culture: note the addition of UserCustomCulture
PS> [CultureInfo]::new("no-su_ch").CultureTypes
SpecificCultures, UserCustomCulture
```

---

Given the above, **at least the following behaviors should be fixed**:

* Remove the nonsensical `InstalledWin32Cultures` flag that is set for _all_ `CultureInfo.CultureTypes`  in Core _on Windows_ ; e.g.:

```powershell
# UNIX: 
PSonUnix> [CultureInfo]::new("no-su_ch").CultureTypes
SpecificCultures, UserCustomCulture

# WINDOWS: !! InstalledWin32Cultures is unexpectedly added
PSonWin> [CultureInfo]::new("no-su_ch").CultureTypes
SpecificCultures, InstalledWin32Cultures, UserCustomCulture
```

* Make `CultureInfo.GetCultures(CultureTypes.UserCustomCulture)` and `CultureInfo.GetCultures(CultureTypes.ReplacementCultures)` output _nothing_ consistently, as is already the case on _Unix_-like platforms.

   * As stated, the current full framework behavior would still make sense even in Core _on Windows_, where _persistently installed_ custom cultures (both supplemental and replacement) can be located with these calls - but perhaps there are _technical_ reasons why that can't be provided in Core.

  * If that behavior cannot be provided, make the calls output _nothing_ on Windows too - that is certainly more sensible than defaulting to _all_ cultures, i.e., treating `UserCustomCulture` and `ReplacementCultures` the same as `AllCultures`.

**Comment by tarekgh** [comment]
> separate issue - let me say that the documentation needs updating

feel free to open an issue in the https://github.com/dotnet/docs 

> Note that we're not just talking about passing those flags to .GetCultures(), but also what the individual CultureInfo instances report in their .CultureTypes property, so you can infer their specific type.

Yes, consider CultureTypes is deprecated too. For .Net core, and we are not touching it in Windows Full framework either. we are keeping it just for legacy and app compat support.

> And you're saying that on Unix-like platforms the concepts of supplemental and replacement cultures (collectively: custom cultures) do not apply altogether?

I am not saying that. what I am saying is Linux/MAC work differently. custom cultures on Windows is different than other OS's. there is a way to create your own locale on Linux too.

> Remove the nonsensical InstalledWin32Cultures flag that is set for all CultureInfo.CultureTypes in Core on Windows ; e.g.:

Nope, InstalledWin32Cultures is still correct in that context. such cultures are really Win32 installed on the system and it has LCID 0x1000 which make it marked as custom. that is what Windows team decided to do.

**Comment by mklement0** [comment]
> feel free to open an issue in the https://github.com/dotnet/docs

Good idea - please see https://github.com/dotnet/docs/issues/7635; I'm pointing to this issue from there, but perhaps you can clarify the points discussed in a more systematic way there.

**Comment by mklement0** [comment]
> Yes, consider CultureTypes is deprecated too.

So even the `NeutralCultures` and `SpecificCultures` flags should be ignored? Again, they do seem to carry meaning currently, across platforms.

> Nope, InstalledWin32Cultures is still correct in that context.

An ad-hoc created culture such as `no-su_ch` or `zz-YY` is clearly not _installed_, so this flag is confusing - not least because the full framework behaves differently.

If `.CultureTypes` is deprecated as a whole, then I suppose it is ultimately irrelevant what it contains, but _omitting_ flags that are deprecated makes more sense to me.

As it stands, the presence of the `InstalledWin32Cultures` is useless noise at best (given that it is reported for _all_ cultures), and a source of confusion at worst.

**Comment by mklement0** [comment]
> I am not saying that. what I am saying is Linux/MAC work differently. custom cultures on Windows is different than other OS's. there is a way to create your own locale on Linux too.

Good to know - that's worth documenting too; again, if you could provide guidance in the [linked docs issue](https://github.com/dotnet/docs/issues/7635), that would be helpful.

**Comment by mklement0** [comment]
Aside from the `.CultureTypes` issue, that leaves us with  `CultureInfo.GetCultures(CultureTypes.UserCustomCulture)` and `CultureInfo.GetCultures(CultureTypes.ReplacementCultures)` exhibiting inconsistent behavior across platforms, which should be fixed, given that these arguments are deprecated, as you state.
